Adding native qconcat

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32335 Benchmarks for quantized LSTM (static)
* #32074 Static quantization for LSTM (front)
* #31934 Static quantization for LSTM (back)
* #31851 Quantized sigmoid function
* **#32252 Adding native qconcat**

Differential Revision: [D19422889](https://our.internmc.facebook.com/intern/diff/D19422889)

Commit:
Adding native qconcat

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 2cb14183:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=2cb1418343bd9d4fecc273c5bbf2bb214335c2ec).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4262841) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86750718) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86750718))
<details>
<summary>
<code>Jan 16 01:29:16 torch_xla/csrc/aten_xla_type_default.cpp:9339:103: error: no member named 'DispatchKey' in namespace 'at'; did you mean 'Dispatcher'?</code>
</summary>

```
Jan 16 01:29:16 /opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/dispatch/Dispatcher.h:35:18: note: 'Dispatcher' declared here 
Jan 16 01:29:16 class CAFFE2_API Dispatcher final { 
Jan 16 01:29:16                  ^ 
Jan 16 01:29:16 torch_xla/csrc/aten_xla_type_default.cpp:9336:138: error: no member named 'DispatchKey' in namespace 'at'; did you mean 'Dispatcher'? 
Jan 16 01:29:16       .impl_unboxedOnlyKernel<at::Tensor(const at::Tensor &, const at::Tensor &, bool), &AtenXlaTypeDefault::_cholesky_solve_helper>(at::DispatchKey::XLATensorId) 
Jan 16 01:29:16                                                                                                                                      ~~~~^~~~~~~~~~~ 
Jan 16 01:29:16                                                                                                                                          Dispatcher 
Jan 16 01:29:16 /opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/dispatch/Dispatcher.h:35:18: note: 'Dispatcher' declared here 
Jan 16 01:29:16 class CAFFE2_API Dispatcher final { 
Jan 16 01:29:16                  ^ 
Jan 16 01:29:16 torch_xla/csrc/aten_xla_type_default.cpp:9339:103: error: no member named 'DispatchKey' in namespace 'at'; did you mean 'Dispatcher'? 
Jan 16 01:29:16       .impl_unboxedOnlyKernel<at::Tensor &(at::Tensor &, bool), &AtenXlaTypeDefault::_coalesced_>(at::DispatchKey::XLATensorId) 
Jan 16 01:29:16                                                                                                   ~~~~^~~~~~~~~~~ 
Jan 16 01:29:16                                                                                                       Dispatcher 
Jan 16 01:29:16 /opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/dispatch/Dispatcher.h:35:18: note: 'Dispatcher' declared here 
Jan 16 01:29:16 class CAFFE2_API Dispatcher final { 
Jan 16 01:29:16                  ^ 
Jan 16 01:29:16 fatal error: too many errors emitted, stopping now [-ferror-limit=] 
Jan 16 01:29:16 20 errors generated. 
packages/torch/include/torch/csrc/api/include -I/opt/conda/lib/python3.6/site-packages/torch/include/TH -I/opt/conda/lib/python3.6/site-packages/torch/include/THC -I/opt/conda/include/python3.6m -c torch_xla/csrc/layout_manager.cpp -o build/temp.linux-x86_64-3.6/torch_xla/csrc/layout_manager.o -std=c++14 -Wno-sign-compare -Wno-deprecated-declarations -Wno-return-type -Wno-macro-redefined -Wno-return-std-move -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=_XLAC -D_GLIBCXX_USE_CXX11_ABI=1 
/lib/python3.6/site-packages/torch/include/torch/csrc/api/include -I/opt/conda/lib/python3.6/site-packages/torch/include/TH -I/opt/conda/lib/python3.6/site-packages/torch/include/THC -I/opt/conda/include/python3.6m -c torch_xla/csrc/view.cpp -o build/temp.linux-x86_64-3.6/torch_xla/csrc/view.o -std=c++14 -Wno-sign-compare -Wno-deprecated-declarations -Wno-return-type -Wno-macro-redefined -Wno-return-std-move -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=_XLAC -D_GLIBCXX_USE_CXX11_ABI=1 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qconcat.cpp:
```diff
@@ -122,5 +124,19 @@ static auto registry =
+
+Tensor quantized_cat(TensorList qxs, int64_t dim) {
+  double _scale = qxs[0].q_scale();
+  int64_t _zero_point = qxs[0].q_zero_point();
```
check for qscheme?
and check that scale and zero_point match for all inputs?

z-a-f replied:
I don't think we need to check as we have a dequant/quant steps. Should I still have the check?

jerryzh168 replied:
`q_scale` is only valid for `PerTensorAffine` quantized tensor

jerryzh168 replied:
we have check in `quantized_cat_impl` but I think it's probably too late if you are calling `q_scale` here

Commit:
Update on "Adding native qconcat"


Differential Revision: [D19422889](https://our.internmc.facebook.com/intern/diff/D19422889)

Commit:
Update on "Adding native qconcat"


Differential Revision: [D19422889](https://our.internmc.facebook.com/intern/diff/D19422889)

Commit:
Update on "Adding native qconcat"


Differential Revision: [D19422889](https://our.internmc.facebook.com/intern/diff/D19422889)

Commit:
Update on "Adding native qconcat"


Differential Revision: [D19422889](https://our.internmc.facebook.com/intern/diff/D19422889)

Commit:
Update on "Adding native qconcat"


Differential Revision: [D19422889](https://our.internmc.facebook.com/intern/diff/D19422889)

supriyar reviewed (approved):
LGTM.

Commit:
Update on "Adding native qconcat"


Differential Revision: [D19422889](https://our.internmc.facebook.com/intern/diff/D19422889)
